{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "# mitofree.py\n",
    "\n",
    "import argparse, functools, os\n",
    "from mitoassembly import mitoassembly\n",
    "\n",
    "print = functools.partial(print, flush=True) #All \"print\" functions have flush=True as default. This way, its contents are not buffered, being instead flushed to the standard output. With this, stdout and stderr redirection works like a charm...\n",
    "\n",
    "\n",
    "# def create_folders(new_working_dir): ##Creates folder for each dataset and changes the working directory\n",
    "#         print(\"New working directory is '%s'\\n\" % (new_working_dir))\n",
    "#         try:\n",
    "#             output_mkdir = os.system(\"mkdir -p %s\" % (new_working_dir))\n",
    "#             os.chdir(new_working_dir)\n",
    "#             return True\n",
    "#         except re.search(\".*Permission denied$\", output_mkdir):\n",
    "#             print(\"Could not create folder %s. Permission denied.\" % (new_working_dir.split(\"/\")[-1]))\n",
    "#             return False\n",
    "#         except:\n",
    "#             print(\"Could not create folder %s\" % (new_working_dir.split(\"/\")[-1]))\n",
    "#             return False\n",
    "\n",
    "# def main_function(sra_list):\n",
    "#     try:\n",
    "#         self.create_folders()\n",
    "#         if self.download_sra_files_prefetch():\n",
    "#             max_read_length = self.highest_read_length()\n",
    "#             self.generate_fastq()\n",
    "#             self.download_seed()\n",
    "#             self.run_NOVOPlasty()\n",
    "#             if self.merge_priority():##Could use this to check if NOVOPlasty assembly has successfully finished and skip this step.\n",
    "#                 print(\"NOVOPlasty assembly succesfully finished!\")\n",
    "#                 self.changeid_pre_mitobim()\n",
    "#                 ##Add while loop that runs mitobim and, if timeout, run generate_fastq with half the read number and then runs mitobim again:\n",
    "#                 #while not run_mitobim(\"largest_contig.fa\", species, name_of_fastq_file):\n",
    "#                 #args.subset = args.subset/2\n",
    "#                 #run_mitobim(\"largest_contig.fa\", species, name_of_fastq_file)\n",
    "#                 self.run_mitobim()\n",
    "#                 last_it = self.last_finalized_iteration()\n",
    "#                 ace = self.mitobim_convert_maf_to_ace(species, last_it)\n",
    "#                 assembly = self.mitobim_ace_to_fasta(ace)\n",
    "#                 if args.savespace:\n",
    "#                     self.remove_assembly_files(name_of_fastq_file)\n",
    "#     except Exception as error:\n",
    "#         print(\"\\nAn error has occurred for this assembly:\\n\\n{}\\n\\nProceeding to the next assembly\\n\".format(error))\n",
    "#         continue\n",
    "#     else:\n",
    "#         pass #PUT ANNOTATION MODULE HERE!!!!!\n",
    "\n",
    "# return(\"All done!\")\n",
    "# ##Add the merge contigs and count contigs here (with its ifs, for readability)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WORKING DIR: /home/gabriel/Dropbox/repos/mitofree/Species1-ERR1306022\n",
      "\n",
      "The file Species1-ERR1306022.sra has already been downloaded. Assembly will proceed normally.\n",
      "\n",
      "Converting ERR1306022 to fastq...\n",
      "Dataset has been converted to fastq succesfully!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gabriel/anaconda3/lib/python3.7/site-packages/Bio/Entrez/__init__.py:606: UserWarning: \n",
      "Email address is not specified.\n",
      "\n",
      "To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "email address with each request.  As an example, if your email address\n",
      "is A.N.Other@example.com, you can specify it as follows:\n",
      "   from Bio import Entrez\n",
      "   Entrez.email = 'A.N.Other@example.com'\n",
      "In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "a user at the email address provided before blocking access to the\n",
      "E-utilities.\n",
      "  E-utilities.\"\"\", UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seed file MK297287.1.seed.fa downloaded succesfully\n",
      "An error has occurred for this assembly: 'int' object is not callable\n",
      "\n",
      "FULL ERROR:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/gabriel/Dropbox/repos/mitofree/mitoassembly.py\", line 56, in main_function\n",
      "    self.create_NOVOPlasty_config()\n",
      "  File \"/home/gabriel/Dropbox/repos/mitofree/mitoassembly.py\", line 155, in create_NOVOPlasty_config\n",
      "    config_out.write(template.read().format(self.prefix, self.kmer, self.maxmemory, self.seed_file, self.max_read_length(), self.fastq_file))\n",
      "TypeError: 'int' object is not callable\n",
      "\n",
      "\n",
      "Proceeding to the next assembly...\n",
      "\n",
      "WORKING DIR: /home/gabriel/Dropbox/repos/mitofree/Species2-RR7295165\n",
      "\n",
      "Downloading RR7295165:\n",
      "\n",
      "The RR7295165 dataset could not be downloaded. Prefetch stderr:\n",
      "warn: nothing found for RR7295165\n",
      "\n",
      "\n",
      "WORKING DIR: /home/gabriel/Dropbox/repos/mitofree/Species3-ERR1306034\n",
      "\n",
      "The file Species3-ERR1306034.sra has already been downloaded. Assembly will proceed normally.\n",
      "\n",
      "Converting ERR1306034 to fastq...\n",
      "Dataset has been converted to fastq succesfully!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gabriel/anaconda3/lib/python3.7/site-packages/Bio/Entrez/__init__.py:606: UserWarning: \n",
      "Email address is not specified.\n",
      "\n",
      "To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "email address with each request.  As an example, if your email address\n",
      "is A.N.Other@example.com, you can specify it as follows:\n",
      "   from Bio import Entrez\n",
      "   Entrez.email = 'A.N.Other@example.com'\n",
      "In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "a user at the email address provided before blocking access to the\n",
      "E-utilities.\n",
      "  E-utilities.\"\"\", UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seed file MK291745.seed.fa downloaded succesfully\n",
      "An error has occurred for this assembly: 'int' object is not callable\n",
      "\n",
      "FULL ERROR:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/gabriel/Dropbox/repos/mitofree/mitoassembly.py\", line 56, in main_function\n",
      "    self.create_NOVOPlasty_config()\n",
      "  File \"/home/gabriel/Dropbox/repos/mitofree/mitoassembly.py\", line 155, in create_NOVOPlasty_config\n",
      "    config_out.write(template.read().format(self.prefix, self.kmer, self.maxmemory, self.seed_file, self.max_read_length(), self.fastq_file))\n",
      "TypeError: 'int' object is not callable\n",
      "\n",
      "\n",
      "Proceeding to the next assembly...\n",
      "\n",
      "WORKING DIR: /home/gabriel/Dropbox/repos/mitofree/Species4-SRR4409513\n",
      "\n",
      "The file Species4-SRR4409513.sra has already been downloaded. Assembly will proceed normally.\n",
      "\n",
      "Converting SRR4409513 to fastq...\n",
      "Dataset has been converted to fastq succesfully!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gabriel/anaconda3/lib/python3.7/site-packages/Bio/Entrez/__init__.py:606: UserWarning: \n",
      "Email address is not specified.\n",
      "\n",
      "To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "email address with each request.  As an example, if your email address\n",
      "is A.N.Other@example.com, you can specify it as follows:\n",
      "   from Bio import Entrez\n",
      "   Entrez.email = 'A.N.Other@example.com'\n",
      "In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "a user at the email address provided before blocking access to the\n",
      "E-utilities.\n",
      "  E-utilities.\"\"\", UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seed file MK291678.seed.fa downloaded succesfully\n",
      "An error has occurred for this assembly: 'int' object is not callable\n",
      "\n",
      "FULL ERROR:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/gabriel/Dropbox/repos/mitofree/mitoassembly.py\", line 56, in main_function\n",
      "    self.create_NOVOPlasty_config()\n",
      "  File \"/home/gabriel/Dropbox/repos/mitofree/mitoassembly.py\", line 155, in create_NOVOPlasty_config\n",
      "    config_out.write(template.read().format(self.prefix, self.kmer, self.maxmemory, self.seed_file, self.max_read_length(), self.fastq_file))\n",
      "TypeError: 'int' object is not callable\n",
      "\n",
      "\n",
      "Proceeding to the next assembly...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# def getArgs():\n",
    "# \tparser = argparse.ArgumentParser(description=\"Downloads sra NGS data and assembles mitochondrial contigs using NOVOPlasty and MITObim\")\n",
    "# \tparser.add_argument(\"-S\", \"--savespace\", action=\"store_true\", default=False, help=\"Automatically removes residual assembly files such as fastq and mitobim iterations\")\n",
    "# \tparser.add_argument(\"-M\", \"--maxmemory\", type=int, metavar=\"\", default=0, help=\"Limit of RAM usage for NOVOPlasty. Default: no limit\")\n",
    "# \tparser.add_argument(\"-K\", \"--kmer\", type=int, metavar=\"\", default=39, help=\"K-mer used in NOVOPlasty assembly. Default: 39\")\n",
    "# \tparser.add_argument(\"-s\", \"--subset\", type=int, metavar=\"\", default=50000000, help=\"Max number of reads used in the assembly process. Default: 50 million reads\")\n",
    "# \t#parser.add_argument(\"-P\", \"--parallel\", type=int, metavar=\"\", default=1, help=\"Number of parallel assemblies. Default: 1\")\n",
    "# \tparser.add_argument(\"-T\", \"--timeout\", type=int, metavar=\"\", default=24, help=\"Custom timeout for MITObim, in hours. Default: 24h\")\n",
    "# \tparser.add_argument(\"filename\", type=str, metavar=\"FILENAME\", help=\"Path to file with multiple accessions (one per line)\")\n",
    "# \treturn parser.parse_args()\n",
    "\n",
    "\n",
    "\n",
    "class args:\n",
    "    filename = \"input_file_example.txt\"    \n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    #args = getArgs()\n",
    "    with open(args.filename) as dataset_list:\n",
    "        for dataset in dataset_list:\n",
    "            assembly = mitoassembly(dataset)\n",
    "            assembly.main_function()\n",
    "            #assembly.main_function()\n",
    "            #Do something here\n",
    "            \n",
    "            #os.chdir(\"../\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/gabriel/Dropbox/repos/mitofree'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/home/gabriel/Dropbox/repos/mitofree/teste\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "environ({'SHELL': '/bin/bash', 'SESSION_MANAGER': 'local/cocozinho:@/tmp/.ICE-unix/1398,unix/cocozinho:/tmp/.ICE-unix/1398', 'WINDOWID': '71303171', 'COLORTERM': 'truecolor', 'XDG_CONFIG_DIRS': '/etc/xdg', 'XDG_SESSION_PATH': '/org/freedesktop/DisplayManager/Session0', 'XDG_MENU_PREFIX': 'xfce-', 'CONDA_EXE': '/home/gabriel/anaconda3/bin/conda', '_CE_M': '', 'LC_ADDRESS': 'pt_BR.UTF-8', 'LC_NAME': 'pt_BR.UTF-8', 'SSH_AUTH_SOCK': '/tmp/ssh-U0DIdA0T8rJj/agent.1459', 'DESKTOP_SESSION': 'xfce', 'LC_MONETARY': 'pt_BR.UTF-8', 'SSH_AGENT_PID': '1460', 'EDITOR': '/usr/bin/nano', 'GTK_MODULES': 'canberra-gtk-module:canberra-gtk-module', 'XDG_SEAT': 'seat0', 'PWD': '/home/gabriel', 'LOGNAME': 'gabriel', 'XDG_SESSION_DESKTOP': 'xfce', 'QT_QPA_PLATFORMTHEME': 'qt5ct', 'XDG_SESSION_TYPE': 'x11', 'CONDA_PREFIX': '/home/gabriel/anaconda3', 'XAUTHORITY': '/home/gabriel/.Xauthority', 'XDG_GREETER_DATA_DIR': '/var/lib/lightdm-data/gabriel', 'GTK2_RC_FILES': '/home/gabriel/.gtkrc-2.0', 'HOME': '/home/gabriel', 'LC_PAPER': 'pt_BR.UTF-8', 'LANG': 'en_US.utf8', 'LS_COLORS': 'rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=01;05;37;41:mi=01;05;37;41:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.axv=01;35:*.anx=01;35:*.ogv=01;35:*.ogx=01;35:*.pdf=00;32:*.ps=00;32:*.txt=00;32:*.patch=00;32:*.diff=00;32:*.log=00;32:*.tex=00;32:*.doc=00;32:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.axa=00;36:*.oga=00;36:*.spx=00;36:*.xspf=00;36:', 'XDG_CURRENT_DESKTOP': 'XFCE', 'VTE_VERSION': '5802', 'CONDA_PROMPT_MODIFIER': '(base) ', 'XDG_SEAT_PATH': '/org/freedesktop/DisplayManager/Seat0', 'YAOURT_COLORS': 'nb=1:pkg=1:ver=1;32:lver=1;45:installed=1;42:grp=1;34:od=1;41;5:votes=1;44:dsc=0:other=1;35', 'XDG_SESSION_CLASS': 'user', 'LC_IDENTIFICATION': 'pt_BR.UTF-8', 'TERM': 'xterm-color', '_CE_CONDA': '', 'USER': 'gabriel', 'CONDA_SHLVL': '1', 'DISPLAY': ':0.0', 'SHLVL': '1', 'LC_TELEPHONE': 'pt_BR.UTF-8', 'LC_MEASUREMENT': 'pt_BR.UTF-8', 'XDG_VTNR': '7', 'XDG_SESSION_ID': 'c2', 'CONDA_PYTHON_EXE': '/home/gabriel/anaconda3/bin/python', 'MOZ_PLUGIN_PATH': '/usr/lib/mozilla/plugins', 'XDG_RUNTIME_DIR': '/run/user/1000', 'CONDA_DEFAULT_ENV': 'base', 'LC_TIME': 'pt_BR.UTF-8', 'LC_ALL': 'C', 'XDG_DATA_DIRS': '/usr/local/share:/usr/share:/var/lib/snapd/desktop:/usr/share', 'PATH': '/home/gabriel/anaconda3/bin:/home/gabriel/anaconda3/condabin:/home/gabriel/.local/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/var/lib/snapd/snap/bin:/home/gabriel/bioinfo/mira_4.0.2_linux-gnu_x86_64_static/bin:/home/gabriel/bioinfo/phylip-3.697/exe:/home/gabriel/bioinfo/sratoolkit.2.10.0-ubuntu64/bin:/home/gabriel/bioinfo/partitionfinder-2.1.1:/home/gabriel/bioinfo/PhyML-3.1:/home/gabriel/bioinfo', 'GDMSESSION': 'xfce', 'DBUS_SESSION_BUS_ADDRESS': 'unix:path=/run/user/1000/bus', 'MAIL': '/var/spool/mail/gabriel', 'LC_NUMERIC': 'pt_BR.UTF-8', '_': '/home/gabriel/anaconda3/bin/jupyter-lab', 'KERNEL_LAUNCH_TIMEOUT': '40', 'JPY_PARENT_PID': '964', 'CLICOLOR': '1', 'PAGER': 'cat', 'GIT_PAGER': 'cat', 'MPLBACKEND': 'module://ipykernel.pylab.backend_inline'})\n"
     ]
    }
   ],
   "source": [
    "print(os.environ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package Bio.Entrez in Bio:\n",
      "\n",
      "NAME\n",
      "    Bio.Entrez - Provides code to access NCBI over the WWW.\n",
      "\n",
      "DESCRIPTION\n",
      "    The main Entrez web page is available at:\n",
      "    http://www.ncbi.nlm.nih.gov/Entrez/\n",
      "    \n",
      "    Entrez Programming Utilities web page is available at:\n",
      "    http://www.ncbi.nlm.nih.gov/books/NBK25501/\n",
      "    \n",
      "    This module provides a number of functions like ``efetch`` (short for\n",
      "    Entrez Fetch) which will return the data as a handle object. This is\n",
      "    a standard interface used in Python for reading data from a file, or\n",
      "    in this case a remote network connection, and provides methods like\n",
      "    ``.read()`` or offers iteration over the contents line by line. See\n",
      "    also \"What the heck is a handle?\" in the Biopython Tutorial and\n",
      "    Cookbook: http://biopython.org/DIST/docs/tutorial/Tutorial.html\n",
      "    http://biopython.org/DIST/docs/tutorial/Tutorial.pdf\n",
      "    \n",
      "    Unlike a handle to a file on disk from the ``open(filename)`` function,\n",
      "    which has a ``.name`` attribute giving the filename, the handles from\n",
      "    ``Bio.Entrez`` all have a ``.url`` attribute instead giving the URL\n",
      "    used to connect to the NCBI Entrez API.\n",
      "    \n",
      "    All the functions that send requests to the NCBI Entrez API will\n",
      "    automatically respect the NCBI rate limit (of 3 requests per second\n",
      "    without an API key, or 10 requests per second with an API key) and\n",
      "    will automatically retry when encountering transient failures\n",
      "    (i.e. connection failures or HTTP 5XX codes). By default, Biopython\n",
      "    does a maximum of three tries before giving up, and sleeps for 15\n",
      "    seconds between tries. You can tweak these parameters by setting\n",
      "    ``Bio.Entrez.max_tries`` and ``Bio.Entrez.sleep_between_tries``.\n",
      "    \n",
      "    The Entrez module also provides an XML parser which takes a handle\n",
      "    as input.\n",
      "    \n",
      "    Variables:\n",
      "    \n",
      "        - email        Set the Entrez email parameter (default is not set).\n",
      "        - tool         Set the Entrez tool parameter (default is ``biopython``).\n",
      "        - api_key      Personal API key from NCBI. If not set, only 3 queries per\n",
      "                       seconds are allowed. 10 queries per seconds otherwise with a\n",
      "                       valid API key.\n",
      "    \n",
      "    Functions:\n",
      "    \n",
      "        - efetch       Retrieves records in the requested format from a list of one or\n",
      "          more primary IDs or from the user's environment\n",
      "        - epost        Posts a file containing a list of primary IDs for future use in\n",
      "          the user's environment to use with subsequent search strategies\n",
      "        - esearch      Searches and retrieves primary IDs (for use in EFetch, ELink,\n",
      "          and ESummary) and term translations and optionally retains\n",
      "          results for future use in the user's environment.\n",
      "        - elink        Checks for the existence of an external or Related Articles link\n",
      "          from a list of one or more primary IDs.  Retrieves primary IDs\n",
      "          and relevancy scores for links to Entrez databases or Related\n",
      "          Articles;  creates a hyperlink to the primary LinkOut provider\n",
      "          for a specific ID and database, or lists LinkOut URLs\n",
      "          and Attributes for multiple IDs.\n",
      "        - einfo        Provides field index term counts, last update, and available\n",
      "          links for each database.\n",
      "        - esummary     Retrieves document summaries from a list of primary IDs or from\n",
      "          the user's environment.\n",
      "        - egquery      Provides Entrez database counts in XML for a single search\n",
      "          using Global Query.\n",
      "        - espell       Retrieves spelling suggestions.\n",
      "        - ecitmatch    Retrieves PubMed IDs (PMIDs) that correspond to a set of\n",
      "          input citation strings.\n",
      "    \n",
      "        - read         Parses the XML results returned by any of the above functions.\n",
      "          Typical usage is:\n",
      "    \n",
      "              >>> from Bio import Entrez\n",
      "              >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "              >>> handle = Entrez.einfo() # or esearch, efetch, ...\n",
      "              >>> record = Entrez.read(handle)\n",
      "              >>> handle.close()\n",
      "    \n",
      "           where record is now a Python dictionary or list.\n",
      "    \n",
      "        - parse        Parses the XML results returned by those of the above functions\n",
      "          which can return multiple records - such as efetch, esummary\n",
      "          and elink. Typical usage is:\n",
      "    \n",
      "              >>> handle = Entrez.esummary(db=\"pubmed\", id=\"19304878,14630660\", retmode=\"xml\")\n",
      "              >>> records = Entrez.parse(handle)\n",
      "              >>> for record in records:\n",
      "              ...     # each record is a Python dictionary or list.\n",
      "              ...     print(record['Title'])\n",
      "              Biopython: freely available Python tools for computational molecular biology and bioinformatics.\n",
      "              PDB file parser and structure class implemented in Python.\n",
      "              >>> handle.close()\n",
      "    \n",
      "          This function is appropriate only if the XML file contains\n",
      "          multiple records, and is particular useful for large files.\n",
      "    \n",
      "        - _open        Internally used function.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    Parser\n",
      "\n",
      "FUNCTIONS\n",
      "    ecitmatch(**keywds)\n",
      "        Retrieve PMIDs for input citation strings, returned as a handle.\n",
      "        \n",
      "        ECitMatch retrieves PubMed IDs (PMIDs) that correspond to a set of input\n",
      "        citation strings.\n",
      "        \n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ECitMatch\n",
      "        \n",
      "        Return a handle to the results, by default in plain text\n",
      "        \n",
      "        Raises an IOError exception if there's a network error.\n",
      "        \n",
      "        Short example:\n",
      "        \n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> citation_1 = {\"journal_title\": \"proc natl acad sci u s a\",\n",
      "        ...               \"year\": \"1991\", \"volume\": \"88\", \"first_page\": \"3248\",\n",
      "        ...               \"author_name\": \"mann bj\", \"key\": \"citation_1\"}\n",
      "        >>> handle = Entrez.ecitmatch(db=\"pubmed\", bdata=[citation_1])\n",
      "        >>> print(handle.read().strip().split(\"|\"))\n",
      "        ['proc natl acad sci u s a', '1991', '88', '3248', 'mann bj', 'citation_1', '2014248']\n",
      "        >>> handle.close()\n",
      "    \n",
      "    efetch(db, **keywords)\n",
      "        Fetch Entrez results which are returned as a handle.\n",
      "        \n",
      "        EFetch retrieves records in the requested format from a list of one or\n",
      "        more UIs or from user's environment.\n",
      "        \n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch\n",
      "        \n",
      "        Return a handle to the results.\n",
      "        \n",
      "        Raises an IOError exception if there's a network error.\n",
      "        \n",
      "        Short example:\n",
      "        \n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> handle = Entrez.efetch(db=\"nucleotide\", id=\"AY851612\", rettype=\"gb\", retmode=\"text\")\n",
      "        >>> print(handle.readline().strip())\n",
      "        LOCUS       AY851612                 892 bp    DNA     linear   PLN 10-APR-2007\n",
      "        >>> handle.close()\n",
      "        \n",
      "        This will automatically use an HTTP POST rather than HTTP GET if there\n",
      "        are over 200 identifiers as recommended by the NCBI.\n",
      "        \n",
      "        **Warning:** The NCBI changed the default retmode in Feb 2012, so many\n",
      "        databases which previously returned text output now give XML.\n",
      "    \n",
      "    egquery(**keywds)\n",
      "        Provide Entrez database counts for a global search.\n",
      "        \n",
      "        EGQuery provides Entrez database counts in XML for a single search\n",
      "        using Global Query.\n",
      "        \n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EGQuery\n",
      "        \n",
      "        Return a handle to the results in XML format.\n",
      "        \n",
      "        Raises an IOError exception if there's a network error.\n",
      "        \n",
      "        This quick example based on a longer version from the Biopython\n",
      "        Tutorial just checks there are over 60 matches for 'Biopython'\n",
      "        in PubMedCentral:\n",
      "        \n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> handle = Entrez.egquery(term=\"biopython\")\n",
      "        >>> record = Entrez.read(handle)\n",
      "        >>> handle.close()\n",
      "        >>> for row in record[\"eGQueryResult\"]:\n",
      "        ...     if \"pmc\" in row[\"DbName\"]:\n",
      "        ...         print(int(row[\"Count\"]) > 60)\n",
      "        True\n",
      "    \n",
      "    einfo(**keywds)\n",
      "        Return a summary of the Entrez databases as a results handle.\n",
      "        \n",
      "        EInfo provides field names, index term counts, last update, and\n",
      "        available links for each Entrez database.\n",
      "        \n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EInfo\n",
      "        \n",
      "        Return a handle to the results, by default in XML format.\n",
      "        \n",
      "        Raises an IOError exception if there's a network error.\n",
      "        \n",
      "        Short example:\n",
      "        \n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> record = Entrez.read(Entrez.einfo())\n",
      "        >>> 'pubmed' in record['DbList']\n",
      "        True\n",
      "    \n",
      "    elink(**keywds)\n",
      "        Check for linked external articles and return a handle.\n",
      "        \n",
      "        ELink checks for the existence of an external or Related Articles link\n",
      "        from a list of one or more primary IDs;  retrieves IDs and relevancy\n",
      "        scores for links to Entrez databases or Related Articles; creates a\n",
      "        hyperlink to the primary LinkOut provider for a specific ID and\n",
      "        database, or lists LinkOut URLs and attributes for multiple IDs.\n",
      "        \n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ELink\n",
      "        \n",
      "        Return a handle to the results, by default in XML format.\n",
      "        \n",
      "        Raises an IOError exception if there's a network error.\n",
      "        \n",
      "        This example finds articles related to the Biopython application\n",
      "        note's entry in the PubMed database:\n",
      "        \n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> pmid = \"19304878\"\n",
      "        >>> handle = Entrez.elink(dbfrom=\"pubmed\", id=pmid, linkname=\"pubmed_pubmed\")\n",
      "        >>> record = Entrez.read(handle)\n",
      "        >>> handle.close()\n",
      "        >>> print(record[0][\"LinkSetDb\"][0][\"LinkName\"])\n",
      "        pubmed_pubmed\n",
      "        >>> linked = [link[\"Id\"] for link in record[0][\"LinkSetDb\"][0][\"Link\"]]\n",
      "        >>> \"17121776\" in linked\n",
      "        True\n",
      "        \n",
      "        This is explained in much more detail in the Biopython Tutorial.\n",
      "    \n",
      "    epost(db, **keywds)\n",
      "        Post a file of identifiers for future use.\n",
      "        \n",
      "        Posts a file containing a list of UIs for future use in the user's\n",
      "        environment to use with subsequent search strategies.\n",
      "        \n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EPost\n",
      "        \n",
      "        Return a handle to the results.\n",
      "        \n",
      "        Raises an IOError exception if there's a network error.\n",
      "    \n",
      "    esearch(db, term, **keywds)\n",
      "        Run an Entrez search and return a handle to the results.\n",
      "        \n",
      "        ESearch searches and retrieves primary IDs (for use in EFetch, ELink\n",
      "        and ESummary) and term translations, and optionally retains results\n",
      "        for future use in the user's environment.\n",
      "        \n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESearch\n",
      "        \n",
      "        Return a handle to the results which are always in XML format.\n",
      "        \n",
      "        Raises an IOError exception if there's a network error.\n",
      "        \n",
      "        Short example:\n",
      "        \n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> handle = Entrez.esearch(db=\"nucleotide\", retmax=10, term=\"opuntia[ORGN] accD\", idtype=\"acc\")\n",
      "        >>> record = Entrez.read(handle)\n",
      "        >>> handle.close()\n",
      "        >>> int(record[\"Count\"]) >= 2\n",
      "        True\n",
      "        >>> \"EF590893.1\" in record[\"IdList\"]\n",
      "        True\n",
      "        >>> \"EF590892.1\" in record[\"IdList\"]\n",
      "        True\n",
      "    \n",
      "    espell(**keywds)\n",
      "        Retrieve spelling suggestions as a results handle.\n",
      "        \n",
      "        ESpell retrieves spelling suggestions, if available.\n",
      "        \n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESpell\n",
      "        \n",
      "        Return a handle to the results, by default in XML format.\n",
      "        \n",
      "        Raises an IOError exception if there's a network error.\n",
      "        \n",
      "        Short example:\n",
      "        \n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> record = Entrez.read(Entrez.espell(term=\"biopythooon\"))\n",
      "        >>> print(record[\"Query\"])\n",
      "        biopythooon\n",
      "        >>> print(record[\"CorrectedQuery\"])\n",
      "        biopython\n",
      "    \n",
      "    esummary(**keywds)\n",
      "        Retrieve document summaries as a results handle.\n",
      "        \n",
      "        ESummary retrieves document summaries from a list of primary IDs or\n",
      "        from the user's environment.\n",
      "        \n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESummary\n",
      "        \n",
      "        Return a handle to the results, by default in XML format.\n",
      "        \n",
      "        Raises an IOError exception if there's a network error.\n",
      "        \n",
      "        This example discovers more about entry 19923 in the structure\n",
      "        database:\n",
      "        \n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> handle = Entrez.esummary(db=\"structure\", id=\"19923\")\n",
      "        >>> record = Entrez.read(handle)\n",
      "        >>> handle.close()\n",
      "        >>> print(record[0][\"Id\"])\n",
      "        19923\n",
      "        >>> print(record[0][\"PdbDescr\"])\n",
      "        Crystal Structure Of E. Coli Aconitase B\n",
      "    \n",
      "    parse(handle, validate=True, escape=False)\n",
      "        Parse an XML file from the NCBI Entrez Utilities into python objects.\n",
      "        \n",
      "        This function parses an XML file created by NCBI's Entrez Utilities,\n",
      "        returning a multilevel data structure of Python lists and dictionaries.\n",
      "        This function is suitable for XML files that (in Python) can be represented\n",
      "        as a list of individual records. Whereas 'read' reads the complete file\n",
      "        and returns a single Python list, 'parse' is a generator function that\n",
      "        returns the records one by one. This function is therefore particularly\n",
      "        useful for parsing large files.\n",
      "        \n",
      "        Most XML files returned by NCBI's Entrez Utilities can be parsed by\n",
      "        this function, provided its DTD is available. Biopython includes the\n",
      "        DTDs for most commonly used Entrez Utilities.\n",
      "        \n",
      "        If validate is True (default), the parser will validate the XML file\n",
      "        against the DTD, and raise an error if the XML file contains tags that\n",
      "        are not represented in the DTD. If validate is False, the parser will\n",
      "        simply skip such tags.\n",
      "        \n",
      "        If escape is True, all characters that are not valid HTML are replaced\n",
      "        by HTML escape characters to guarantee that the returned strings are\n",
      "        valid HTML fragments. For example, a less-than sign (<) is replaced by\n",
      "        &lt;. If escape is False (default), the string is returned as is.\n",
      "        \n",
      "        Whereas the data structure seems to consist of generic Python lists,\n",
      "        dictionaries, strings, and so on, each of these is actually a class\n",
      "        derived from the base type. This allows us to store the attributes\n",
      "        (if any) of each element in a dictionary my_element.attributes, and\n",
      "        the tag name in my_element.tag.\n",
      "    \n",
      "    read(handle, validate=True, escape=False)\n",
      "        Parse an XML file from the NCBI Entrez Utilities into python objects.\n",
      "        \n",
      "        This function parses an XML file created by NCBI's Entrez Utilities,\n",
      "        returning a multilevel data structure of Python lists and dictionaries.\n",
      "        Most XML files returned by NCBI's Entrez Utilities can be parsed by\n",
      "        this function, provided its DTD is available. Biopython includes the\n",
      "        DTDs for most commonly used Entrez Utilities.\n",
      "        \n",
      "        If validate is True (default), the parser will validate the XML file\n",
      "        against the DTD, and raise an error if the XML file contains tags that\n",
      "        are not represented in the DTD. If validate is False, the parser will\n",
      "        simply skip such tags.\n",
      "        \n",
      "        If escape is True, all characters that are not valid HTML are replaced\n",
      "        by HTML escape characters to guarantee that the returned strings are\n",
      "        valid HTML fragments. For example, a less-than sign (<) is replaced by\n",
      "        &lt;. If escape is False (default), the string is returned as is.\n",
      "        \n",
      "        Whereas the data structure seems to consist of generic Python lists,\n",
      "        dictionaries, strings, and so on, each of these is actually a class\n",
      "        derived from the base type. This allows us to store the attributes\n",
      "        (if any) of each element in a dictionary my_element.attributes, and\n",
      "        the tag name in my_element.tag.\n",
      "\n",
      "DATA\n",
      "    api_key = None\n",
      "    email = None\n",
      "    max_tries = 3\n",
      "    print_function = _Feature((2, 6, 0, 'alpha', 2), (3, 0, 0, 'alpha', 0)...\n",
      "    sleep_between_tries = 15\n",
      "    tool = 'biopython'\n",
      "\n",
      "FILE\n",
      "    /home/gabriel/anaconda3/lib/python3.7/site-packages/Bio/Entrez/__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "help(Entrez)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.efetch()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
